### PR TITLE
Implement NCMEC Upvote/Downvote in reference API

### LINKED ISSUE #1614: [py-tx] Implement NCMEC Upvote/Downvote in reference API
NCMEC has added a new API to "upvote" and "downvote" hash entries.

Add a simple implementation to call that API. Here's the documentation: https://report.cybertip.org/hashsharing/v2/documentation/ (ctrl+f vote).

You won't be able to get NCMEC credentials, but @dcallies can test that the implementation is right by patching your PR and using Meta's test API access. 

The PR should include a unittest that mocks the NCMEC API response, and potentially updating the existing test response to include feedback data: https://github.com/facebook/ThreatExchange/blob/main/python-threatexchange/threatexchange/exchanges/clients/ncmec/tests/test_hash_api.py#L38

### PR DESCRIPTION
**Pull request by aryzle**
Summary
---------

Implement NCMEC Upvote/Downvote in reference API
docs: https://report.cybertip.org/hashsharing/v2/documentation/#working-with-feedback

fixes #1614

Test Plan
---------

TBD

### CODE REVIEW

File: python-threatexchange/threatexchange/exchanges/clients/ncmec/tests/data.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+            <guid>01234567-abcd-0123-4567-012345678900</guid>
+        </reasonIds>
+    </negative>
+</feedbackSubmission>
+""".strip()
+
+UPDATE_FEEDBACK_RESULT_XML = """
+<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
+<submissionResult xmlns="https://hashsharing.ncmec.org/hashsharing/v2">
+  <!-- What this returns is not documented -->
```
Comment (aryzle) [inline python-threatexchange/threatexchange/exchanges/clients/ncmec/tests/data.py:200]: @Dcallies feedback submission has no response documented [here](https://report.cybertip.org/hashsharing/v2/documentation/#working-with-feedback)

File: python-threatexchange/threatexchange/exchanges/clients/ncmec/hash_api.py [hunk 2]
```diff
@@ -19,14 +19,15 @@
...
-from requests.packages.urllib3.util.retry import Retry
+from urllib3.util.retry import Retry
 from threatexchange.exchanges.clients.utils.common import TimeoutHTTPAdapter
 
 
 _DATE_FORMAT_STR = "%Y-%m-%dT%H:%M:%SZ"
 _DEFAULT_ELE = ET.Element("")
 
 T = t.TypeVar("T")
+FEEDBACK = t.List[t.Dict[str, str]]
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/exchanges/clients/ncmec/hash_api.py:30]: blocking: ALL_CAPS usually denotes a constant. In this case, your type is simple enough that you should just inline it.

File: python-threatexchange/threatexchange/exchanges/clients/ncmec/hash_api.py [hunk 3]
```diff
@@ -131,6 +132,7 @@ class NCMECEntryUpdate:
     deleted: bool
     classification: t.Optional[str]
     fingerprints: t.Dict[str, str]
+    feedback: t.Optional[FEEDBACK]
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/exchanges/clients/ncmec/hash_api.py:135]: blocking: We should not use optional here, because empty dictionary expresses the same thing

File: python-threatexchange/threatexchange/exchanges/clients/ncmec/hash_api.py [hunk 4]
```diff
@@ -148,6 +150,18 @@ def from_xml(cls, xml: _XMLWrapper) -> "NCMECEntryUpdate":
             fingerprints={
                 x.tag: x.text for x in xml.maybe("fingerprints") if x.has_text
             },
+            feedback=[
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/exchanges/clients/ncmec/hash_api.py:153]: blocking; Your type says dict, your code says list!

File: python-threatexchange/threatexchange/exchanges/clients/ncmec/hash_api.py [hunk 5]
```diff
@@ -401,6 +432,14 @@ def get_entries_iter(
             has_more = bool(next_)
             yield result
 
+    # TODO: split into 2, submit upvote and downvote
+    def submit_feedback(self, entry_id: str, is_good: bool) -> GetEntriesResponse:
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/exchanges/clients/ncmec/hash_api.py:436]: unsure: This API also needs the ESP ID, no?

File: python-threatexchange/threatexchange/exchanges/clients/ncmec/hash_api.py [hunk 6]
```diff
@@ -401,6 +432,14 @@ def get_entries_iter(
             has_more = bool(next_)
             yield result
 
+    # TODO: split into 2, submit upvote and downvote
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/exchanges/clients/ncmec/hash_api.py:435]: blocking; Yup

send_upvote(...):

send_downvote(...):

File: python-threatexchange/threatexchange/exchanges/clients/ncmec/tests/data.py [hunk 7]
```diff
@@ -0,0 +1,202 @@
+STATUS_XML = """
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/exchanges/clients/ncmec/tests/data.py:1]: This is fine, copy paste the copyright header too if you have moment

File: python-threatexchange/threatexchange/exchanges/clients/ncmec/tests/test_hash_api.py [hunk 8]
```diff
@@ -11,166 +10,17 @@
...
-
-NEXT_UNESCAPED = (
-    "/v2/entries?from=2017-10-20T00%3A00%3A00.000Z"
-    "&to=2017-10-30T00%3A00%3A00.000Z&start=2001&size=1000&max=3000"
-)
-
-NEXT_UNESCAPED2 = (
-    "/v2/entries?from=2017-10-20T00%3A00%3A00.000Z"
-    "&to=2017-10-30T00%3A00%3A00.000Z&start=3001&size=1000&max=4000"
+from threatexchange.exchanges.clients.ncmec.tests.data import (
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/exchanges/clients/ncmec/tests/test_hash_api.py:13]: ignorable: Consider importing by module when you are importing many functions.

**Review by Dcallies — CHANGES_REQUESTED** [review]
Looks like a good start. Please

File: python-threatexchange/threatexchange/exchanges/clients/ncmec/hash_api.py (hunk 4 above)
Comment (aryzle) [reply python-threatexchange/threatexchange/exchanges/clients/ncmec/hash_api.py:165]: type is `t.List[t.Dict[str, str]]`

File: python-threatexchange/threatexchange/exchanges/clients/ncmec/hash_api.py [hunk 9]
```diff
@@ -148,6 +162,18 @@ def from_xml(cls, xml: _XMLWrapper) -> "NCMECEntryUpdate":
...
             },
+            feedback=[
+                {
+                    "type": x.tag,  # "affirmativeFeedback" or "negativeFeedback"
+                    "members": x.maybe(
+                        "members"  # "timestamp", "member.id", "member.name"
+                    ),
+                    "reasons": x.maybe(
+                        "reasons"
+                    ),  # "reason" with "guid", "name", "type" | "members"
```
Comment (aryzle) [inline python-threatexchange/threatexchange/exchanges/clients/ncmec/hash_api.py:173]: working on expanding `members` and `reasons` here

**Comment by aryzle** [comment]
@Dcallies this should be testable against the NCMEC API now, I just have a few minor things to iron out

File: python-threatexchange/threatexchange/exchanges/clients/ncmec/hash_api.py [hunk 10]
```diff
@@ -261,15 +343,19 @@ def __init__(
         username: str,
         password: str,
         environment: NCMECEnvironment,
+        member_id: t.Optional[str] = None,
+        reasons_map: t.Dict[str, t.List[t.Dict[str, str]]] = {},
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/exchanges/clients/ncmec/hash_api.py:347]: blocking: We won't know this at client construction time, and can poll it afterwards.

File: python-threatexchange/threatexchange/exchanges/clients/ncmec/hash_api.py [hunk 11]
```diff
@@ -261,15 +343,19 @@ def __init__(
...
     ) -> None:
         assert is_valid_user_pass(username, password)
         self.username = username
         self.password = password
         self._base_url = environment.value
+        self.member_id = member_id
+        self.reasons_map = reasons_map or {}
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/exchanges/clients/ncmec/hash_api.py:354]: Hmm this `or {}` is fairly subtle. I'm assuming we are doing it because the default is `{}` which is a shared instance across any object, and the `or {}` will replace it with a new instance per call, but having it be optional and default to  `None` is more easy to prove doesn't have a bug.

File: python-threatexchange/threatexchange/exchanges/clients/ncmec/hash_api.py [hunk 12]
```diff
@@ -401,6 +535,55 @@ def get_entries_iter(
...
 
+    def submit_feedback(
+        self,
+        entry_id: str,
+        feedback_type: NCMECFeedbackType,
+        affirmative: bool,
+        reason_id: t.Optional[str] = None,
+    ) -> GetEntriesResponse:
+        if not affirmative and not reason_id:
+            raise ValueError("Negative feedback must have a reason_id")
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/exchanges/clients/ncmec/hash_api.py:546]: (note to self): If reason map only has one item in it, choose that item. Also test NCMEC API to make sure it won't accept empty reason, which should be our pref otherwise.

File: python-threatexchange/threatexchange/exchanges/clients/ncmec/hash_api.py [hunk 13]
```diff
@@ -123,6 +123,19 @@ class NCMECEntryType(Enum):
     video = "video"
 
 
+@unique
+class NCMECFeedbackType(Enum):
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/exchanges/clients/ncmec/hash_api.py:127]: (note to self): Add note that these are also the entry types

File: python-threatexchange/threatexchange/exchanges/clients/ncmec/hash_api.py [hunk 14]
```diff
@@ -131,6 +144,7 @@ class NCMECEntryUpdate:
     deleted: bool
     classification: t.Optional[str]
     fingerprints: t.Dict[str, str]
+    feedback: t.List[t.Dict[str, t.Any]]
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/exchanges/clients/ncmec/hash_api.py:147]: (note to self): Consider whether it makes sense to use the enum here, or whether to leave as string in all cases

File: python-threatexchange/threatexchange/exchanges/clients/ncmec/hash_api.py [hunk 15]
```diff
@@ -148,6 +162,36 @@ def from_xml(cls, xml: _XMLWrapper) -> "NCMECEntryUpdate":
...
+                [
+                    {
+                        "sentiment": x.tag,  # "affirmativeFeedback" or "negativeFeedback"
+                        "type": x.str("type"),
+                        "latest_feedback_time": x.str("lastUpdateTimestamp"),
+                        "members": [
+                            {"id": m.str("id"), "name": m.text}
+                            for m in x.maybe("members")
+                            if m.has_text
+                        ],
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/exchanges/clients/ncmec/hash_api.py:175]: (note to self): Should this be dict instead of synthesizing a a dict? What about a dedicated class?

File: python-threatexchange/threatexchange/exchanges/clients/ncmec/hash_api.py [hunk 16]
```diff
@@ -215,11 +259,49 @@ def estimated_entries_in_range(self) -> int:
...
+    @classmethod
+    def from_xml(cls, xml: _XMLWrapper) -> "GetFeedbackReasonsResponse":
+        reasons = []
+        for reason in xml.maybe("availableFeedbackReasons"):
+            reasons.append(
+                {
+                    "guid": reason.str("guid"),
+                    "name": reason.str("name"),
+                    "type": reason.str("type"),
+                }
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/exchanges/clients/ncmec/hash_api.py:294]: (note to self): Use a dedicated class here instead of dict

File: python-threatexchange/threatexchange/exchanges/clients/ncmec/hash_api.py [hunk 17]
```diff
@@ -401,6 +535,55 @@ def get_entries_iter(
             has_more = bool(next_)
             yield result
 
+    def submit_feedback(
+        self,
+        entry_id: str,
+        feedback_type: NCMECFeedbackType,
+        affirmative: bool,
+        reason_id: t.Optional[str] = None,
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/exchanges/clients/ncmec/hash_api.py:543]: (note to self): Is a reason string better to use here than the guid?

**Review by Dcallies — COMMENTED** [review]
Thanks for all the hard work @aryzle ! We'll take it from here and test it against the real API and finish it off! You can consider this one done!
